### PR TITLE
KoLeo regularizer

### PR DESCRIPTION
**Pull request by gabrieltseng**
Implements the KoLeo regularizer

Screenshot 2025-04-01 at 12 50 10 PM

Example run: https://wandb.ai/eai-ai2/helios-debug/runs/vj4k4bgo

### CODE REVIEW

File: helios/train/loss.py [hunk 1]
```diff
@@ -517,6 +517,73 @@ def compute(
...
+                _online encoder outputs_, not decoder outputs.
+            targets: Unused, and only kept for consistency.
+            **kwargs: Additional keyword arguments.
+
+        Returns:
+            The computed loss value.
+        """
+        all_preds, all_masks = predictions.flatten_tokens_and_masks()
+        online_encodings = all_preds[all_masks == MaskValue.ONLINE_ENCODER.value]
+
```
Comment (gabrieltseng) [inline helios/train/loss.py:596]: Currently this is computed against **all patch embeddings**. Could this be too expensive? It runs at the moment but I believe Dinov2 does it against "backbone embeddings" - one thing we could do is compute an instance average before this regularizer is applied.

Comment (gabrieltseng) [reply helios/train/loss.py:596]: This OOMs on Galileo with a microbatch size of 128 but works with 64
https://wandb.ai/eai-ai2/helios-debug/runs/8t4ktbhi

Screenshot 2025-04-01 at 2 04 35 PM

Comment (Hgherzog) [reply helios/train/loss.py:596]: I think an instance average makes sense and aligns more closely with what is done in dino and aligns with using a mean pooling of the features

File: helios/train/train_module/train_module.py [hunk 2]
```diff
@@ -484,3 +488,12 @@ def eval_batch(
...
         raise NotImplementedError("eval batch not implemented")
+
+    def add_regularizer_to_loss(
+        self, loss: torch.Tensor, latent: TokensAndMasks
+    ) -> torch.Tensor:
+        """If a regularizer is present, add it to the loss."""
+        regularizer = getattr(self, "regularizer", None)
+        if regularizer is None:
+            return loss
+        return loss + regularizer.compute(latent, None)
```
Comment (Hgherzog) [inline helios/train/train_module/train_module.py]: Suggestion: log the regularizer metric separetly as well

File: helios/train/train_module/latent_mim.py [hunk 3]
```diff
@@ -166,6 +169,13 @@ def __init__(
         self.token_exit_cfg = token_exit_cfg
         self.base_loss = loss_config.build()
         self.masking_strategy = masking_config.build()
+        self.regularizer = (
+            regularizer_config.build() if regularizer_config is not None else None
+        )
+
+        self.metric_name = self.base_loss.name
```
Comment (Hgherzog) [inline helios/train/train_module/latent_mim.py]: Suggestion: metric name is unclear, perhaps total_loss_name?

File: helios/train/loss.py [hunk 4]
```diff
@@ -517,6 +517,73 @@ def compute(
...
+        """
+        # parwise dot products (= inverse distance)
+        dots = torch.mm(x, x.t())
+        n = x.shape[0]
+        dots.view(-1)[:: (n + 1)].fill_(-1)  # Trick to fill diagonal with -1
+        # max inner prod -> min distance
+        _, indices = torch.max(dots, dim=1)
+        return indices
+
+    def compute(
```
Comment (Hgherzog) [inline helios/train/loss.py:575]: we should add some way to weight the regularizer here.  See page 29 https://arxiv.org/pdf/2304.07193 DinoV2 uses a weight of 0.1

File: helios/train/train_module/mae.py [hunk 5]
```diff
@@ -160,6 +162,12 @@ def __init__(
         self.token_exit_cfg = token_exit_cfg
         self.base_loss = loss_config.build()
         self.masking_strategy = masking_config.build()
+        self.regularizer = (
+            regularizer_config.build() if regularizer_config is not None else None
+        )
+        self.metric_name = self.base_loss.name
+        if self.regularizer is not None:
```
Comment (Hgherzog) [inline helios/train/train_module/mae.py:169]: Suggestion: Another option instead of having to bind the names here is to either make a meta loss class that allows for arbitrary weighted combinations of losses. This also might be more amenable to the case where we want to add multiple regularizing terms

File: helios/train/loss.py (hunk 1 above)
Comment (gabrieltseng) [reply helios/train/loss.py:596]: updated

File: helios/train/loss.py (hunk 4 above)
Comment (gabrieltseng) [reply helios/train/loss.py:575]: added

File: helios/train/train_module/mae.py (hunk 5 above)
Comment (gabrieltseng) [reply helios/train/train_module/mae.py:169]: I agree and this is what I originally intended to do, but a challenge is that different losses receive different inputs (e.g. this regularizer is computed against the latents), and its generally a bit module specific (e.g. Galileo's losses being branch specific)

I guess we could have all loss functions receive *three* inputs (latents, preds, targets) but this seems like it would be good to do when we have an example of at least one other loss which requires the latents to establish the pattern

Comment (gabrieltseng) [reply helios/train/train_module/mae.py:169]: Actually on reflection this is possible but I'd prefer to do it in a separate PR to keep the changes contained. 

Issue created: https://github.com/allenai/helios/issues/175

File: helios/train/train_module/latent_mim.py (hunk 3 above)
Comment (gabrieltseng) [reply helios/train/train_module/latent_mim.py]: updated

File: helios/train/train_module/mae.py [hunk 6]
```diff
@@ -223,7 +231,7 @@ def train_batch(
                 loss.backward()
 
         self.trainer.record_metric(
-            f"train/{self.base_loss.name}",
+            f"train/{self.total_loss_name}",
             total_batch_loss / get_world_size(self.dp_process_group),
             ReduceType.mean,
```
Comment (gabrieltseng) [inline helios/train/train_module/mae.py:250]: MAE is the only module to divide the `total_batch_loss` by `get_world_size(self.dp_process_group)` - probably the other modules should be doing this or MAE shouldn't?

File: helios/train/train_module/galileo.py [hunk 7]
```diff
@@ -212,13 +221,9 @@ def train_batch(
         # Set the model to train mode
         self.model.train()
 
-        # This is a clear loss buffer
-        if not hasattr(self, "total_batch_loss"):
-            self.total_batch_loss = torch.zeros(1, device=self.device)
-        else:
-            self.total_batch_loss.fill_(0.0)
```
Comment (gabrieltseng) [inline helios/train/train_module/galileo.py:219]: I assume this is outdated code and safe to remove? no other train module has it.

File: helios/train/train_module/train_module.py (hunk 2 above)
Comment (gabrieltseng) [reply helios/train/train_module/train_module.py]: okay - added but I'd appreciate if you double checked my implementation (plus two other comments)

File: helios/train/train_module/mae.py (hunk 6 above)
Comment (Hgherzog) [reply helios/train/train_module/mae.py:250]: Yes MAE shouldn't Reduce_Type.mean already averages the metric across the gpus

Comment (Hgherzog) [reply helios/train/train_module/mae.py:250]: by yes I mean that MAE shouldn't

File: helios/train/train_module/galileo.py (hunk 7 above)
Comment (Hgherzog) [reply helios/train/train_module/galileo.py:219]: The idea here was to remove a host device sync, it ended up not really making a big difference though because we have to sync again immediately after. Option to either remove or add methods to helios train module. This would be like a set and clear loss buffers method (I think the trainer already calls clear loss buffers)
